nixos/modules/services/networking/pihole-ftl.nix

```
@@ -0,0 +1,344 @@
+              default = "";
+            };
+          };
+          default = { };
+          description = mdDoc "Pi-hole adlist definition";
+        };
+      in
+      mkOption {
+        type = with types; listOf attrs;
+        description = mdDoc "URLs of blocklists to use";
+        default = [ ];
+        example = [{ url = "https://raw.githubusercontent.com/StevenBlack/hosts/master/hosts"; }];
+      };
+
+    user = mkOption {
+      type = types.str;
+      default = "pihole";
+      description = mdDoc "User to run the service as.";
+    };
+
+    group = mkOption {
+      type = types.str;
+      default = "pihole";
+      description = mdDoc "Group to run the service as.";
+    };
+  };
+
+  config = mkIf cfg.enable {
+
+    assertions = [
+      {
+        assertion = !config.services.dnsmasq.enable;
+        message = "pihole-ftl conflicts with dnsmasq. Please disable one of them";
+      }
+    ];
+
+    systemd.tmpfiles.rules = [
+      "d ${cfg.stateDirectory} 0700 ${cfg.user} ${cfg.group} - -"
+      "f ${logFile} 0700 ${cfg.user} ${cfg.group} - -"
+      "f ${debugFile} 0700 ${cfg.user} ${cfg.group} - -"
+    ];
+
+    systemd.services = {
+      pihole-ftl-setup = {
+        description = "Pi-hole FTL setup";
+        after = [ "network.target" ];
+        serviceConfig = let ftlCfg = config.systemd.services.pihole-ftl; in {
+          Type = "oneshot";
+          User = ftlCfg.serviceConfig.User;
+          Group = ftlCfg.serviceConfig.Group;
+        };
+        script =
+          let
+            concatMapLines = f: list: builtins.concatStringsSep "\n    " (map f list);
+            escapeSQL = builtins.replaceStrings [ "'" ] [ "''" ];
+          in
+          ''
+            set -euo pipefail
+            pihole=${cfg.pihole}/bin/pihole
+
+            # If the database doesn't exist, it needs to be created with gravity.sh
+            if [ ! -f ${cfg.stateDirectory}/gravity.db ]; then
+              DONT_RESTART_FTL=1 $pihole -g
+            fi
+
+            ${concatMapLines
+              (list: ''$pihole -a adlist add \
+                ${strings.escapeShellArg (escapeSQL list.url)} \
+                ${strings.escapeShellArg (escapeSQL list.comment)}'')
+              cfg.adlists}
+
+            # Run gravity.sh to load any new adlists
+            DONT_RESTART_FTL=1 $pihole -g
+          '';
+      };
+
+      pihole-ftl = {
+        description = "Pi-hole FTL";
+
+        after = [ "network.target" "pihole-ftl-setup.service" ];
+        requires = [ "pihole-ftl-setup.service" ];
+        wantedBy = [ "multi-user.target" ];
+
+        serviceConfig = {
+          Type = "simple";
+          User = cfg.user;
+          Group = cfg.group;
+          AmbientCapabilities = [
+            "CAP_NET_BIND_SERVICE"
+            "CAP_NET_RAW"
+            "CAP_NET_ADMIN"
+            "CAP_SYS_NICE"
+          ];
+          ExecStart = "${cfg.package}/bin/pihole-FTL no-daemon -- --conf-file=${validatedDnsmasqConf}";
```

```suggestion
          ExecStart = "${lib.getExe cfg.package} no-daemon -- --conf-file=${validatedDnsmasqConf}";
```
you might need to set meta.mainProgram for this to work